pkg/apis/discovery/types.go

```diff
@@ -0,0 +1,124 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// EndpointSlice represents a set of endpoints that implement a service
+type EndpointSlice struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// +optional
+	metav1.ObjectMeta
+	// This field specifies the list of ports associated with each endpoint in the EndpointSlice
+	// Each EndpointPort must have a unique port name.
+	// +optional
+	// +listType=atomic
+	Ports []EndpointPort
+	// This field specifies the type of targets associated with each endpoint in the EndpointSlice
+	// Default is IP.
+	// +optional
+	TargetType *TargetType
+	// The set of endpoints backing this slice
+	// +listType=atomic
+	Endpoints []Endpoint
+}
+
+// Endpoint represents a set of backends implementing a service
+type Endpoint struct {
+	// Required: Targets of the endpoint. Must contain at least one target.
+	// Different consumers (e.g. kube-proxy) handle different types of
+	// targets in the context of its own capabilities.
+	// +listType=set
+	Targets []string
+	// Required: the conditions of the endpoint.
+	Conditions EndpointConditions
+	// The stated hostname of this endpoint. This field may be used by consumers
+	// of endpoint to distinguish endpoints from each other (e.g. in DNS names).
+	// Multiple endpoints which use the same hostname should be considered fungible (e.g. multiple A values in DNS).
+	// +optional
+	Hostname *string
+	// Reference to object providing the endpoint.
+	// +optional
+	TargetRef *api.ObjectReference
+	// Topology can contain arbitrary topology information associated with the
+	// endpoint.
+	// Key/value pairs contained in topology must conform with the label format.
+	// https://kubernetes.io/docs/concepts/overview/working-with-objects/labels
+	// Topology may include, but is not limited to the following well known keys:
+	// kubernetes.io/hostname: the value indicates the hostname of the node
+	// where the endpoint is located. This should match the corresponding
+	// node label.
+	// topology.kubernetes.io/zone: the value indicates the zone where the
+	// endpoint is located. This should match the corresponding node label.
+	// topology.kubernetes.io/region: the value indicates the region where the
+	// endpoint is located. This should match the corresponding node label.
+	// +optional
+	Topology map[string]string
+}
+
+// EndpointConditions represents the current condition of an endpoint
+type EndpointConditions struct {
+	// Ready indicates if the endpoint is ready to serve traffic
+	Ready bool
```

Comment:
From what I can tell, `don't know` has a tendency to cover lots of different states. (e.g. endpoint has not been health checked vs. endpoint has not been programmed) Hence, making it hard to consume. The consumer is likely to fall back to ready/notReady

Reply:
What I am afraid of is that there could be cases where the caller legit does not know, and we're forcing them to say yes.  Alternatively we could say nil means "don't know" and callers should treat it as a tentative yes

Reply:
or not - not super locked into this one

Reply:
it just felt like a field I shouldn't have to specify if I don't know.

Reply:
You are right. Maybe making the Ready field an optional pointer?

Reply:
I'm truly torn on this.  If we make it optional, what will kube-proxy do in case of `nil`?  Maybe it's not such a hardship to require users to set this, given that most users are controllers...  Or we could default it to true, so only "clever" controllers which add not-yet-ready addresses need to do anything.

What do you think?  You guys have as much voice in this as I do.

Reply:
Just a recap: A producer of endpointSlice can leave a condition nil if it "does-not-apply" or "does-not-know". Then the condition meant "unknown" or "unspecified" to the consumer. Then it is up to the consumer to interpret the condition (the consumer may not care about the condition). I feel like we can only enforce kube-proxy/kube-dns to interpret "not specified" as ready. 

That sort of mean we can only document, "consumer **should** interpreted as yes" instead of "must".  WDYT?

Reply:
Yeah, that's pretty much my (B), but more accurate.